modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/RuntimeMessageHandlerImpl.java

```diff
@@ -86,56 +89,43 @@ public void beforeDeactivation(ComponentManager mgr) {
     }
 
     protected void changeStep(ComponentManagerStep step) {
-        messages.removeIf(msg -> step.equals(msg.getStep()));
-        this.step = step;
+        this.currentStep = step;
     }
 
-    /**
-     * @since 9.10
-     */
-    protected static class Message {
-
-        protected final ComponentManagerStep step;
-
-        protected final Level level;
-
-        protected final String message;
-
-        public Message(ComponentManagerStep step, Level level, String message) {
-            this.step = step;
-            this.level = level;
-            this.message = message;
-        }
-
-        public ComponentManagerStep getStep() {
-            return step;
-        }
-
-        public Level getLevel() {
-            return level;
-        }
-
-        public String getMessage() {
-            return message;
-        }
-
+    @Override
+    public void addMessage(RuntimeMessage message) {
+        messages.add(message);
     }
 
-    /**
-     * @since 9.10
-     */
-    protected enum ComponentManagerStep {
-
-        ACTIVATING,
-
-        STARTING,
+    @Override
+    public void addMessage(Level level, String message, String source) {
+        addMessage(new RuntimeMessage(currentStep, level, message, source));
+    }
 
-        RUNNING,
+    protected Predicate<RuntimeMessage> getFinalPredicate(Predicate<RuntimeMessage> givenPredicate) {
+        return givenPredicate == null ? m -> true : givenPredicate;
+    }
 
-        STOPPING,
+    @Override
+    public List<RuntimeMessage> getMessages(Predicate<RuntimeMessage> predicate) {
+        final Predicate<RuntimeMessage> p = predicate == null ? m -> true : predicate;
```

Comment:
Can you remove the null argument possibility?

Reply:
See above comment (same reason)

Reply:
Removed the null possibility